client/network/src/service/tests.rs

```diff
@@ -272,3 +272,72 @@ fn notifications_state_consistent() {
 		}
 	});
 }
+
+#[test]
+fn lots_of_incoming_peers_works() {
+	let listen_addr = config::build_multiaddr![Memory(rand::random::<u64>())];
+
+	let (main_node, _) = build_test_full_node(config::NetworkConfiguration {
+		notifications_protocols: vec![(ENGINE_ID, From::from(&b"/foo"[..]))],
+		listen_addresses: vec![listen_addr.clone()],
+		in_peers: u32::max_value(),
+		transport: config::TransportConfig::MemoryOnly,
+		.. config::NetworkConfiguration::new_local()
+	});
+
+	let main_node_peer_id = main_node.local_peer_id().clone();
+
+	// We spawn background tasks and push them in this `Vec`. They will all be waited upon before
+	// this test ends.
+	let mut background_tasks_to_wait = Vec::new();
+
+	for _ in 0..256 {
+		let main_node_peer_id = main_node_peer_id.clone();
+
+		let (_dialing_node, mut event_stream) = build_test_full_node(config::NetworkConfiguration {
+			notifications_protocols: vec![(ENGINE_ID, From::from(&b"/foo"[..]))],
+			listen_addresses: vec![],
+			reserved_nodes: vec![config::MultiaddrWithPeerId {
+				multiaddr: listen_addr.clone(),
+				peer_id: main_node_peer_id.clone(),
+			}],
+			transport: config::TransportConfig::MemoryOnly,
+			.. config::NetworkConfiguration::new_local()
+		});
+
+		background_tasks_to_wait.push(async_std::task::spawn(async move {
+			// Create a dummy timer that will "never" fire, and that will be overwritten when we
+			// actually need the timer. Using an Option would be technically cleaner, but it would
+			// make the code below way more complicated.
+			let mut timer = futures_timer::Delay::new(Duration::from_secs(3600 * 24 * 7)).fuse();
+
+			loop {
+				futures::select! {
+					_ = timer => {
+						// Test succeeds when timer fires.
+						return;
+					}
+					ev = event_stream.next().fuse() => {
```

Comment:
This needs to happen outside the loop, cf. https://docs.rs/futures/0.3.5/futures/macro.select.html.

Reply:
I believe that the documentation warns about that in order to avoid polling the future after completion. Here calling `next()` recreates a new future every single time.

If we call `event_stream.next().fuse()` outside of the loop, then we also have to do `fut = event_stream.next().fuse()` inside of the loop, otherwise we will only ever get the first event.

Reply:
What I had in mind was `let mut event_stream = event_stream.fuse();` outside the loop and `event_stream.next()` in line 320.